spark/src/main/java/org/apache/iceberg/spark/source/SparkPlannerUtil.java

```diff
@@ -0,0 +1,231 @@
+    return metadataTableName(table, table.toString(), type);
+  }
+
+  private static String metadataTableName(Table table, String tableName, MetadataTableType type) {
+    if (tableName.contains("/")) {
+      return tableName + "#" + type;
+    } else if (tableName.startsWith("hadoop.")) {
+      // for HadoopCatalog tables, use the table location to load the metadata table
+      // because IcebergCatalog uses HiveCatalog when the table is identified by name
+      return table.location() + "#" + type;
+    } else if (tableName.startsWith("hive.")) {
+      // HiveCatalog prepend a logical name which we need to drop for Spark 2.4
+      return tableName.replaceFirst("hive\\.", "") + "." + type;
+    } else {
+      return tableName + "." + type;
+    }
+  }
+
+  private static CloseableIterable<FileScanTask> planDataTableScanFiles(
+      SparkSession spark,
+      DataTableScan scan,
+      Snapshot snapshot) {
+
+    //TODO Currently this approach reads all manifests, no manifest filtering - Maybe through pushdowns or table read
+    // options?
+    //Create a dataframe of all DataFile entries
+    String dataFilesMetadataTable = metadataTableName(scan.table(), MetadataTableType.FILES);
+    Dataset<Row> dataFileRows =
+        spark.read()
+            .format("iceberg")
+            .load(dataFilesMetadataTable);
+
+    Types.StructType partitionStruct = DataFile.getType(scan.table().spec().partitionType());
+    StructType sparkSchema = dataFileRows.schema();
+
+    Dataset<SparkDataFile> dataFiles =
+        dataFileRows.mapPartitions((Iterator<Row> it) -> {
+          //Wrap is mutable here and not thread safe
+              SparkDataFile container = new SparkDataFile(partitionStruct, sparkSchema);
+              return Streams.stream(it).map(container::wrap).iterator();
+              },
+            Encoders.javaSerialization(SparkDataFile.class));
+
+
+    final Expression scanFilter = scan.filter();
+    final boolean ignoreResiduals = scan.shouldIgnoreResiduals();
+    final boolean isCaseSensitive = scan.isCaseSensitive();
+
+    Map<Integer, PartitionSpec> specsById = scan.table().specs();
+    JavaSparkContext jsc = JavaSparkContext.fromSparkContext(spark.sparkContext());
+
+    /*
+    Since we aren't filtering entries on manifest read, filter file entries using Partition information
+    In the normal path this happens with the manifest reader's entries code, filtering elements before we
+    get to this stage.
+     */
+
+    //Build cache of partition evaluators
+    ImmutableMap.Builder<Integer, Evaluator> evalMapBuilder = new ImmutableMap.Builder<>();
+    specsById.entrySet().stream().forEach( entry ->
+      evalMapBuilder.put(entry.getKey(),
+          new Evaluator(entry.getValue().partitionType(),
+              Projections.inclusive(entry.getValue(), isCaseSensitive).project(scanFilter))));
+
+    Map<Integer, Evaluator> partitionEvaluatorsById = evalMapBuilder.build();
```

Comment:
If we go this route, we would need to add metrics evaluators as well.

Reply:
If we go this route, would it be possible to make these metrics appear in the spark ui or in the history server ui? Possibly we consider metrics that convert to the native framework's accumulators etc that will print in the UI.

I'm huge on metrics.

Reply:
@kbendick I believe @aokolnychyi was referring to the file metrics (column min/max values etc ...) Which I didn't apply here :) 

I have added them back in the new commits so we are properly filtering on metrics as well. As for Visibility in the UI currently the planning job will just appear as another job, I do think we should probably log some stats from the planning stage as well but we can look into that in a future pr.